python/cucim/src/cucim/core/operations/intensity/normalize.py

```diff
@@ -0,0 +1,135 @@
+
+def normalize_data(
+    img: Any,
+    norm_constant: float,
+    min_value: float,
+    max_value: float,
+    type: int = 1
+) -> Any:
+    """
+    Apply intensity normalization to the input array.
+    Normalize intensities to the range of [0, norm_constant].
+
+    Parameters
+    ----------
+    img : channel first, cupy.ndarray or numpy.ndarray
+        Input data of shape (C, H, W). Can also batch process input of shape
+        (N, C, H, W). Can be a numpy.ndarray or cupy.ndarray.
+    norm_constant: float
+        Normalization range of the input data. [0, norm_constant]
+    min_value : float
+        Minimum intensity value in input data.
+    max_value : float
+        Maximum intensity value in input data.
+    type : int
+        Type of normalization.
+        Supported types are
+            range based : 1
+            atangent based : 2
+
+    Returns
+    -------
+    out : cupy.ndarray or numpy.ndarray
+        Output data. Same dimensions and type as input.
+
+    Raises
+    ------
+    TypeError
+        If input 'img' is not cupy.ndarray or numpy.ndarray
+    ValueError
+        If input original intensity min and max are same
+    ValueError
+        If incorrect normalization type is invoked
+
+    Examples
+    --------
+    >>> import cucim.core.operations.intensity as its
+    >>> # input is channel first 3d array
+    >>> output_array = its.normalize_data(input_arr,
+                                          10, 0 , 255)
+    """
+    try:
+        if max_value - min_value == 0.0:
+            raise ValueError("Minimum and Maximum intensity \
+                              same in input data")
+
+        if type not in [1, 2]:
+            raise ValueError("Incorrect normalization type. \
+                              Supported types are: \
+                                  range based: 1,\
+                                  atangent based: 2")
+
+        to_cupy = False
+
+        if isinstance(img, np.ndarray):
+            to_cupy = True
+            cupy_img = cupy.asarray(img, dtype=cupy.float32, order='C')
+        elif not isinstance(img, cupy.ndarray):
+            raise TypeError("img must be a cupy.ndarray or numpy.ndarray")
+        else:
+            cupy_img = cupy.ascontiguousarray(img)
+
+        if cupy_img.dtype != cupy.float32:
+            if cupy.can_cast(img.dtype, cupy.float32) is False:
+                raise ValueError(
+                    "Cannot safely cast type {cupy_img.dtype.name} to \
+                    'float32'"
+                )
+            else:
+                cupy_img = cupy_img.astype(cupy.float32)
+
+        normalize = CUDA_KERNELS.get_function("normalize_data_by_range")
+
+        if type == 2:
+            normalize = CUDA_KERNELS.get_function("normalize_data_by_atan")
+
+        value_range = max_value - min_value
+        norm_factor = norm_constant / value_range
+
+        sh = img.shape
+        total_size = np.prod(sh)
```

Comment:
```suggestion
        total_size = math.prod(img.shape)
```

`math.prod` was introduced in Python 3.8 and has much less overhead than `np.prod` for small tuples.

example:
```
shape = (1024, 768)
%timeit math.prod(shape)
85.1 ns ± 3.1 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

%timeit np.prod(shape)
3.13 µs ± 38 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```